scripts/data/src/org/akvo/gae/remoteapi/DeleteFormCompletely.java

```diff
@@ -0,0 +1,139 @@
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.TimeZone;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+
+/*
+ * - Delete a form with all it's groups, questions and options
+ */
+public class DeleteFormCompletely implements Process {
+
+    @Override
+    public void execute(DatastoreService ds, String[] args) throws Exception {
+
+        boolean doit = false;
+        Long surveyId = null;
+        final List<Key> toBeRemoved = new ArrayList<>();
+
+        for (int i = 0; i < args.length; i++) {
+            //System.out.printf("#Argument %d: %s\n", i, args[i]);
+            if (args[i].equalsIgnoreCase("--doit")) {
+                doit = true;
+            } else {
+                surveyId = Long.parseLong(args[i]);
+            }
+        }
+        if (surveyId == null) {
+            System.out.println("No Form id parameter given!");
+            return;
+        }
+
+        DateFormat df = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS");
+        df.setTimeZone(TimeZone.getTimeZone("UT"));
+
+        System.out.println("#Processing form " + surveyId);
+        if (!doit) {
+            System.out.println(" (report only)");
+        }
+
+        try { //May be linked entities around, even if the form is not
+            Entity form = ds.get(KeyFactory.createKey("Survey", surveyId));
+            toBeRemoved.add(form.getKey());
+        }
+        catch (EntityNotFoundException e) {
+            System.out.println("Form not found: " + surveyId);
+        }
+
+        final Query query1 = new Query("QuestionGroup").setFilter(new Query.FilterPredicate("surveyId", FilterOperator.EQUAL, surveyId));
+        final PreparedQuery pq = ds.prepare(query1);
+
+        for (Entity qg : pq.asIterable(FetchOptions.Builder.withChunkSize(500))) {
+            toBeRemoved.add(qg.getKey());
+        }
+
+        final Query query2 = new Query("Question").setFilter(new Query.FilterPredicate("surveyId", FilterOperator.EQUAL, surveyId));
+        final PreparedQuery pq2 = ds.prepare(query2);
+
+        for (Entity q : pq2.asIterable(FetchOptions.Builder.withChunkSize(500))) {
+            if ("OPTION".equals(q.getProperty("type"))) {
+                toBeRemoved.addAll(optionsOfQuestion(ds, q.getKey().getId()));
+            }
+            toBeRemoved.add(q.getKey());
+        }
+
+
+        if (doit) {
+            System.out.println("Deleting " + toBeRemoved.size() + " entities.");
+            ds.delete(toBeRemoved);
```

Comment:
We have a utility function DataUtils.batchDelete() that deletes in batches within limitations of data store queries.